### ISSUE TITLE
WARNorgDomains.ts Match of WEBAPP_URL with ALLOWED_HOSTNAME failed

### ISSUE THREAD

**Issue by mabdelfattah** [calcom/cal.com#12201]
Latest cal version on Docker.

Server logs:
```
WARNorgDomains.ts Match of WEBAPP_URL with ALLOWED_HOSTNAME failed {
   WEBAPP_URL: 'https://cal.mydomain.com',
   ALLOWED_HOSTNAMES: [
       [
           'cal.mydomain.com'
    ]
  ]
}
```

.env File
```
ALLOWED_HOSTNAMES='["cal.mydomain.com"]'
NEXT_PUBLIC_WEBAPP_URL=https://cal.mydomain.com
```

UPDATED:
I updated the .env File to the following, and I can no more see the warning
```
ALLOWED_HOSTNAMES='"mydomain.com"'
NEXT_PUBLIC_WEBAPP_URL=https://cal.mydomain.com
```

**Comment by jabertech** [comment]
I noticed the same failed match on my domain, but when I use the following:
`ALLOWED_HOSTNAMES='"mydomain.com"'
NEXT_PUBLIC_WEBAPP_URL=https://cal.mydomain.com`
I get a 404 error:
![image](https://github.com/calcom/cal.com/assets/138768746/4b8f6030-895e-4c20-bc27-db19e1cbd6bc)

**Comment by jabertech** [comment]
I noticed the same failed match on my domain, but when I use the following:
`ALLOWED_HOSTNAMES='"mydomain.com"'
NEXT_PUBLIC_WEBAPP_URL=https://cal.mydomain.com`
I get a 404 error:
![image](https://github.com/calcom/cal.com/assets/138768746/4b8f6030-895e-4c20-bc27-db19e1cbd6bc)

**Comment by mturilin** [comment]
I also see this error

**Comment by leguennecj** [comment]
same error here

**Comment by drajabr** [comment]
Thanks for your update, this error no longer appear.

However now, I get redirect to http://localhost:3000 after login

**Comment by Krislunde** [comment]
I'm having the same issue when trying to visit my public page. 

<img width="912" alt="Screenshot 2024-02-18 at 00 48 06" src="https://github.com/calcom/cal.com/assets/5624240/da29db9a-4f6d-4102-83d0-c00b46452663">
<img width="1001" alt="Screenshot 2024-02-18 at 00 54 12" src="https://github.com/calcom/cal.com/assets/5624240/9e1e0db3-07cb-4a2f-8204-6d9bc08de024">

Allowed hostnames is properly defined in .env:
`ALLOWED_HOSTNAMES='"cal.com","cal.dev","cal-staging.com","cal.community","cal.local:3050","localhost:3050","book.mydomain.com.au","mydomain.com"'`

**Comment by Krislunde** [comment]
I'm having the same issue when trying to visit my public page.

<img width="912" alt="Screenshot 2024-02-18 at 00 48 06" src="https://github.com/calcom/cal.com/assets/5624240/da29db9a-4f6d-4102-83d0-c00b46452663">
<img width="1001" alt="Screenshot 2024-02-18 at 00 54 12" src="https://github.com/calcom/cal.com/assets/5624240/9e1e0db3-07cb-4a2f-8204-6d9bc08de024">

Allowed hostnames is properly defined in .env:
`ALLOWED_HOSTNAMES='"cal.com","cal.dev","cal-staging.com","cal.community","cal.local:3050","localhost:3050","book.mydomain.com.au","mydomain.com"'`

**Comment by lubomirblazekcz** [comment]
There might be a problem if the hostname is under a subdomain, see https://github.com/calcom/cal.com/blob/b38807cb0f9cc024f4b614d1cb631245bb95117e/packages/features/ee/organizations/lib/orgDomains.ts#L38

subdomain will never match here becuase of the dot

**Comment by Krislunde** [comment]
> There might be a problem if the hostname is under a subdomain

That seems like a likely culprit. I was indeed trying to host under book.mydomain.com

**Comment by olivierlambert** [comment]
I have the same problem with the same situation (subdomain)

**Comment by dantosXD** [comment]
I am also having the same issues on two deployments all under book.mydomain.com

**Comment by Krislunde** [comment]
@zomars Is this a known issue that is being worked on?

**Comment by andr-ec** [comment]
@zomars @PeerRich Should we reopen this issue? If it's something separate I can create a new issue.

**Comment by axsuul** [comment]
I am also hosting under a subdomain and having this issue. Deploying to a subdomain is pretty popular in the self-hosting community so hope to see this addressed soon!

Thanks for the great software.

**Comment by SamAmann** [comment]
I'm joigning the rank of the people having the same issue, under the same condition apparently (self-hosted on a subdomain).
I've tried many:
`WEBAPP_URL: 'https://cal.domain.tld'`

```
ALLOWED_HOSTNAMES: [
@calcom/web:start:     [
@calcom/web:start:       'https://cal.domain.tld',
@calcom/web:start:       'cal.domain.tld',
@calcom/web:start:       'cal.com',
@calcom/web:start:       'cal.dev',
@calcom/web:start:       'cal-staging.com',
@calcom/web:start:       'cal.community',
@calcom/web:start:       'cal.local:3000',
@calcom/web:start:       'localhost:3000'
@calcom/web:start:     ]
@calcom/web:start:   ]
```
as per the .env.example provided.

All resulting in the same error:
`WARNorgDomains.ts Match of WEBAPP_URL with ALLOWED_HOSTNAME failed`

**Comment by drajabr** [comment]
> I'm joigning the rank of the people having the same issue, under the same condition apparently (self-hosted on a subdomain). I've tried many: `WEBAPP_URL: 'https://cal.domain.tld'`

This is the worst project I tinkered with for self-hosting, I already ditched it. It's probably meant to be anti-self-hosted. Sad thing is this behavior only affect poor enthusiasts, as I believe someone with enough money can hire a dev to fix it for their own benefit.

**Comment by siddharthalodha** [comment]
Same bot as well. Same issues.

**Comment by hariombalhara** [comment]
First of all sorry about the delay here, somehow I missed this ticket 🙏

It seems like a configuration and expectations issue to me. Cal.com supports organization with subdomains only at the moment.

`ALLOWED_HOSTNAMES` holds a list of domains, on the subdomains of which the booking links and dashboard would work.

So, if ALLOWED_HOSTNAMES='"cal.com", "cal.dev"' then,
`acme.cal.com` is a valid organization booking domain and WEBAPP_URL can be set to `http://app.cal.com`
`dunder.cal.dev` is also a valid organization subdomain and WEBAPP_URL can be set to `http://app.cal.dev`



If ALLOWED_HOSTNAMES='"cal.domain.tld"" then
WEBAPP_URL should be app.cal.domain.tld and acme.cal.domain.tld, dunder.cal.domain.tld, xxxx.cal.domain.tld are valid organization domain.


Also, to get organizations working please follow https://github.com/calcom/cal.com/blob/main/packages/features/ee/organizations/README.md#app-setup


Though I realize that we need to improve the DX here with better documentation and simplifying things more. Expect improvements soon !!

**Comment by RauchenwaldC** [comment]
> So, if ALLOWED_HOSTNAMES='"cal.com", "cal.dev"' then, `acme.cal.com` is a valid organization booking domain and WEBAPP_URL can be set to `http://app.cal.com` `dunder.cal.dev` is also a valid organization subdomain and WEBAPP_URL can be set to `http://app.cal.dev`
> 
> If ALLOWED_HOSTNAMES='"cal.domain.tld"" then WEBAPP_URL should be app.cal.domain.tld and acme.cal.domain.tld, dunder.cal.domain.tld, xxxx.cal.domain.tld are valid domains.

Thanks for the reply. If I understand this correctly then the following shouldn't show up?

```
@calcom/web:start: 08:47:32:454 WARNorgDomains.ts Match of WEBAPP_URL with ALLOWED_HOSTNAME failed {
@calcom/web:start:   WEBAPP_URL: 'https://calendar.convertain.com',
@calcom/web:start:   ALLOWED_HOSTNAMES: [
@calcom/web:start:     [
@calcom/web:start:       'convertain.com',
@calcom/web:start:       'convertain.dev'
@calcom/web:start:     ]
@calcom/web:start:   ]
@calcom/web:start: }
```

If I set `NEXT_PUBLIC_WEBAPP_URL` just to _calendar.convertain.com_ without the HTTPS in the Docker settings then I don't get the app to run at all because of...
```
+ scripts/replace-placeholder.sh http://localhost:3000 calendar.convertain.com
Replacing all statically built instances of http://localhost:3000 with calendar.convertain.com.
```

**Comment by KastanDay** [comment]
@hariombalhara In the Organization setup instructions, we have to enable Organizations in the web app before we can use them. But we can't access the web app because organizations aren't enabled. 

Therefore, I can't host Cal on a subdomain from the start. It's a classic catch-22: I need organizations, but can't enable them. Can you confirm this interpretation or offer a work around? Thanks so much! I deeply appreciate your work.

**Comment by JackPala** [comment]
I'll add to this, I have two instances one with a reverse proxy at localhost that does not work and has the same error everyone else is describing

However I have another setup with a reverse proxy on a different server, both of which use caddy. 

Oddy enough, the one with the DIFFERENT server running the reverse proxy, appears to work with a subdomain and https without fail.

**Comment by keithwillcode** [comment]
@RauchenwaldC Based on what you've posted, it indeed seems like the warning should be gone.

We are working on improvements to orgs set ups, domain config, Dockerization, etc. at the moment so will report back here with updates over the coming days/weeks. Appreciate your all's patience and understanding.

**Comment by alexisljn** [comment]
I get rid of this issue by using top level domain instead of subdomain. 

I know this is not sustainable but if you absolutely need a functional setup, it works.

**Comment by webexpr-jdevesvrotte** [comment]
Thx @alexisljn!

**Comment by cjangrist** [comment]
sounds like using TLD is the workaround, will try it soon. The login redirect to localhost:3000 gets users stuck because on prod that's obviously not available

**Comment by Fiser12** [comment]
I am facing this same issue

**Comment by ad1196** [comment]
the error is still there, impossible to resolve despite several approaches

**Comment by Rustymage** [comment]
I am in the same boat - looking around, this appears to be an issue from 2023 with little hope of resolution.

```
@calcom/web:start: 08:01:12:548 [WARN] orgDomains.ts Match of WEBAPP_URL with ALLOWED_HOSTNAME failed {
@calcom/web:start:   WEBAPP_URL: 'https://URL.COM',
@calcom/web:start:   ALLOWED_HOSTNAMES: []
@calcom/web:start: }
```

**Comment by jxs714** [comment]
> Latest cal version on Docker.
> 
> Server logs:
> 
> ```
> WARNorgDomains.ts Match of WEBAPP_URL with ALLOWED_HOSTNAME failed {
>    WEBAPP_URL: 'https://cal.mydomain.com',
>    ALLOWED_HOSTNAMES: [
>        [
>            'cal.mydomain.com'
>     ]
>   ]
> }
> ```
> 
> .env File
> 
> ```
> ALLOWED_HOSTNAMES='["cal.mydomain.com"]'
> NEXT_PUBLIC_WEBAPP_URL=https://cal.mydomain.com
> ```
> 
> UPDATED: I updated the .env File to the following, and I can no more see the warning
> 
> ```
> ALLOWED_HOSTNAMES='"mydomain.com"'
> NEXT_PUBLIC_WEBAPP_URL=https://cal.mydomain.com
> ```
> 
> [CAL-3415](https://linear.app/calcom/issue/CAL-3415/warnorgdomainsts-match-of-webapp-url-with-allowed-hostname-failed)

Good find. Dont understand why the allowed_hostname variable is not mentioned anywhere in the documentation....

**Comment by paulux84** [comment]
i continue to have this error also after this settings:

```
ALLOWED_HOSTNAMES='"mydomain.com"'
NEXT_PUBLIC_WEBAPP_URL=https://cal.mydomain.com
```

my deployment is inside kubernetes

**Comment by jxs714** [comment]
> i continue to have this error also after this settings:
> 
> ```
> ALLOWED_HOSTNAMES='"mydomain.com"'
> NEXT_PUBLIC_WEBAPP_URL=https://cal.mydomain.com
> ```
> 
> my deployment is inside kubernetes

Doesn't seem like the developers care at all to fix these kind of issues that shouldn't exist to begin with. Seems like they are focusing more on their commercial product and may be trying to push everyone to just purchase their subscription or self hosted license... First time I see a open source product treated this way with a reputable company...

**Comment by Billos** [comment]
Same problem here...

**Comment by frenchcharly** [comment]
So, if I haven't missed something else : you can't self-host on a subdomain without a fix here, right?

Wouldn't this be solved by having 2 variables : an optional (that defaults to an empty string or 'www') prefix one and the actual domain (with its TLD)?

I'm trying to love cal.com, but so many things seem to have been overlooked, it's a shame.

**Comment by PeerRich** [comment]
> I'm trying to love cal.com, but so many things seem to have been overlooked, it's a shame.

its open source, send patch, will merge

**Comment by drajabr** [comment]
> or adjusting the business logic to account for subdomains ?

It already does.

> While our [ultimate ](https://cal.com/ultimate)level plan offers extensive white-labeling support, inclusive of setting up your own Cal.com subdomain.

But apparently for paid plans and "only?" under their cal.com domain. 

But that's a fair argument from @PeerRich its open source, send a patch, and if it doesn't harm their business model they will merge it, but this will, so its more likely to go under "not quiet sure what benefits / negative side effects this could cause" or even "Oh Unit test fail" thing then go stall forever. 

But hey, please let us know if you did it @frenchcharly , even if not merged, we "amateur self hosters not backed by multi-million  $ company" can use your fork :) I tried to look for the most useful forks in these 9.5k forks, obviously I couldn't look on them all, but the first 100 most useful  didn't address this particular issue.

**Comment by PeerRich** [comment]
this is not a change that would affect our business and we have merged many changes in the past that promote self-hosting

we want people to have a great self-hosting experience, but our team is small and we need to keep focussed

a lot of docker work falls behind. we just hired an agency and will deploy many hundreds of thousands to make the docker experience seamless

**Comment by asadriko** [comment]
Hello everyone! 

I believe this might also be related to the `NEXT_PUBLIC_SINGLE_ORG_SLUG` variable. 

If you could clarify the usage of the `NEXT_PUBLIC_SINGLE_ORG_SLUG` and `ALLOWED_HOSTNAMES` variables, that would be really helpful and would allow us to use them correctly. 

I was able to get it working by setting the variables as follows: 
```
ALLOWED_HOSTNAMES='"cal.mydomain.com"'
NEXT_PUBLIC_SINGLE_ORG_SLUG= 

```
However, I only figured this out through trial and error. I also noticed that using multiple domains caused an error, for example: 
```
ALLOWED_HOSTNAMES='"cal.mydomain.com","mydomain.com"'
NEXT_PUBLIC_SINGLE_ORG_SLUG=
```
If anyone could clarify all the variables used for addresses throughout the application, as well as their relationship and impact, that would be greatly appreciated.

Thank you!

**Comment by lsagetlethias** [comment]
> Hello everyone!
> 
> I believe this might also be related to the `NEXT_PUBLIC_SINGLE_ORG_SLUG` variable.
> 
> If you could clarify the usage of the `NEXT_PUBLIC_SINGLE_ORG_SLUG` and `ALLOWED_HOSTNAMES` variables, that would be really helpful and would allow us to use them correctly.
> 
> I was able to get it working by setting the variables as follows:
> 
> ```
> ALLOWED_HOSTNAMES='"cal.mydomain.com"'
> NEXT_PUBLIC_SINGLE_ORG_SLUG= 
> ```
> 
> However, I only figured this out through trial and error. I also noticed that using multiple domains caused an error, for example:
> 
> ```
> ALLOWED_HOSTNAMES='"cal.mydomain.com","mydomain.com"'
> NEXT_PUBLIC_SINGLE_ORG_SLUG=
> ```
> 
> If anyone could clarify all the variables used for addresses throughout the application, as well as their relationship and impact, that would be greatly appreciated.
> 
> Thank you!

thanks for this! Can you maybe share an obfuscated version of your .env?

I tried like this but no luck yet:
```env
ALLOWED_HOSTNAMES="\"https://calendso.mysubdomain.mydomain.fr\""
NEXT_PUBLIC_SINGLE_ORG_SLUG=""
NEXT_PUBLIC_WEBAPP_URL="https://calendso.mysubdomain.mydomain.fr"
NEXT_PUBLIC_WEBSITE_URL="https://calendso.mysubdomain.mydomain.fr"
NEXTAUTH_URL="https://calendso.mysubdomain.mydomain.fr"
```

**Comment by asadriko** [comment]
> > Hello everyone!
> > I believe this might also be related to the `NEXT_PUBLIC_SINGLE_ORG_SLUG` variable.
> > If you could clarify the usage of the `NEXT_PUBLIC_SINGLE_ORG_SLUG` and `ALLOWED_HOSTNAMES` variables, that would be really helpful and would allow us to use them correctly.
> > I was able to get it working by setting the variables as follows:
> > ```
> > ALLOWED_HOSTNAMES='"cal.mydomain.com"'
> > NEXT_PUBLIC_SINGLE_ORG_SLUG= 
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > However, I only figured this out through trial and error. I also noticed that using multiple domains caused an error, for example:
> > ```
> > ALLOWED_HOSTNAMES='"cal.mydomain.com","mydomain.com"'
> > NEXT_PUBLIC_SINGLE_ORG_SLUG=
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > If anyone could clarify all the variables used for addresses throughout the application, as well as their relationship and impact, that would be greatly appreciated.
> > Thank you!
> 
> thanks for this! Can you maybe share an obfuscated version of your .env?
> 
> I tried like this but no luck yet:
> 
> ALLOWED_HOSTNAMES="\"https://calendso.mysubdomain.mydomain.fr\""
> NEXT_PUBLIC_SINGLE_ORG_SLUG=""
> NEXT_PUBLIC_WEBAPP_URL="https://calendso.mysubdomain.mydomain.fr"
> NEXT_PUBLIC_WEBSITE_URL="https://calendso.mysubdomain.mydomain.fr"
> NEXTAUTH_URL="https://calendso.mysubdomain.mydomain.fr"


```
NEXT_PUBLIC_LICENSE_CONSENT=agree
LICENSE=agree
CALCOM_LICENSE_KEY=cal_live_***************
CAL_SIGNATURE_TOKEN=cal_sk_***********************
CALCOM_PRIVATE_API_ROUTE="https://goblin.cal.com"

NEXT_PUBLIC_WEBAPP_URL=https://example.com
NEXT_PUBLIC_API_V2_URL=http://api.example.com/v2
ALLOWED_HOSTNAMES='"example.com"'

NEXT_PUBLIC_SINGLE_ORG_SLUG=
NEXT_PUBLIC_WEBSITE_URL=https://example.com

NEXT_PUBLIC_WEBSITE_TERMS_URL=https://example.com/terms
NEXT_PUBLIC_WEBSITE_PRIVACY_POLICY_URL=https://example.com/policy

#NEXTAUTH_URL=https://example.com

NEXTAUTH_SECRET=********************************

CALENDSO_ENCRYPTION_KEY=******************************

# JWT_SECRET=********************************

POSTGRES_USER=calcom
POSTGRES_PASSWORD=*******************
POSTGRES_DB=dbcalcom
DATABASE_HOST=your-db-host.amazonaws.com:5432
DATABASE_URL=postgresql://${POSTGRES_USER}:${POSTGRES_PASSWORD}@${DATABASE_HOST}/${POSTGRES_DB}
DATABASE_DIRECT_URL=${DATABASE_URL}
PGSSLMODE='no-verify'
GOOGLE_API_CREDENTIALS={}

CALCOM_TELEMETRY_DISABLED=

MS_GRAPH_CLIENT_ID=
MS_GRAPH_CLIENT_SECRET=

ZOOM_CLIENT_ID=
ZOOM_CLIENT_SECRET=

EMAIL_FROM_NAME=SeuNome
EMAIL_FROM=contato@example.com

EMAIL_SERVER_HOST=email-smtp.us-east-1.amazonaws.com
EMAIL_SERVER_PORT=587
EMAIL_SERVER_USER=AKIA****************
EMAIL_SERVER_PASSWORD=*******************************

NODE_ENV=production
NODE_TLS_REJECT_UNAUTHORIZED=1

API_KEY_PREFIX=cal_

NEXT_PUBLIC_VAPID_PUBLIC_KEY=BD***********************
VAPID_PRIVATE_KEY=********************************

```

**Comment by krumware** [comment]
Depending on your network stack, other things could be at play regarding the ALLOWED_HOSTS. This is typically configured as a way of ensuring that the host header in requests to your services matches what your application should expect.

One way you can test the configuration, is to use cUrl, like `curl --header "Host: mycal.mydomain.com" http://yourserviceaddress`

Where there is some differentiation here, and why there isn't just one variable to rule them all, is that the serverside rendering and even static rendering is using some of the NEXT_ prefixed vars. So there may be some repetition.

The source code does infer that a command delimited list of strings is the desired format. Sometimes your container runtime may wrap the variables in quotes for you, so try passing this as `ALLOWED_HOSTS=test.my.com,other.my.com` (no quotes)
Additionally, change your log level and there are some WARN level logs that might get produced if there's an issue `NEXT_PUBLIC_LOGGER_LEVEL=0`
There is an example above which includes the array brackets, and you can see that the warn message shows an array in an array, so that's a good sign that the syntax is incorrect there. We may also observe extra quotes or escape characters in some of your other examples. Again, this may dependent on your container runtime.

NEXT_PUBLIC_SINGLE_ORG_SLUG syntax is: `NEXT_PUBLIC_SINGLE_ORG_SLUG=booker`
this slug is used in more than the routing, so it isn't directly comparable. The functionality for that is documented in the /cal.com repo readme. It does appear that It may to override the allowed_hosts by accepting {orgSlug.*}, so any site starting with booker.* - I will also re-test this shortly.

**Comment by lsagetlethias** [comment]
The only way it worked for me was by adding "RESERVED_SUBDOMAINS='"calendso"' as my hostname was under 2 subdomains (NEXT_PUBLIC_WEBSITE_URL="https://calendso.mysubdomain.mydomain.fr").
ALLOWED_HOSTS & NEXT_PUBLIC_SINGLE_ORG_SLUG (even with an empty value) were not mandatory, I don't know why and at this point I'm too afraid to ask 😅

TL;DR; my minimal env with 2 levels of subdomains:

```env
RESERVED_SUBDOMAINS="\"calendso\""
NEXT_PUBLIC_WEBAPP_URL="https://calendso.mysubdomain.mydomain.fr/"
NEXT_PUBLIC_WEBSITE_URL="https://calendso.mysubdomain.mydomain.fr/"
NEXTAUTH_URL="https://calendso.mysubdomain.mydomain.fr/"
```

Hope this helps!

**Comment by hariombalhara** [comment]
_NEXT_PUBLIC_SINGLE_ORG_SLUG_ is documented here https://github.com/calcom/cal.com/blob/main/README.md#L490 and it is only needed to be set when you want to use an organization and serve both cal.com dashboard and cal.com booking pages through a single domain.

**Comment by hariombalhara** [comment]
I feel like we should close this issue and have very specific issues for the scenario people are facing, there are many different scenarios in this issue that are already resolved.

**Comment by hariombalhara** [comment]
Did some updates here in the PR for README.
https://github.com/calcom/cal.com/pull/22060

I am closing this ticket. Please open separate tickets with the specific details about your setup if you see a related issue.

**Comment by krumware** [comment]
Thank you!

**Comment by saschafoerster** [comment]
Worked for me, thanks! :)
